### PR TITLE
Bug 1561958 - introduce cacheOverRestarts option

### PR DESCRIPTION
**Pull request by djmitche**
This option serializes the `Run` instance to disk and then skips calling `ConfigureRun` or getting secrets if that serialization file already exists.

I chose to serialize the whole Run rather than just the worker config because I want to maintain support for features like credentials expiration, which require some metadata about the run (such as when the credentials expire).  This will also allow us to support such future features such as polling for deployment ID changes, even after a restart.

There's also a huge refactor in here to try to move the "meat' of the application out of `cmd/start-worker` and into a top-level package.  Go's lack of support for inter-referential packages makes this rather difficult, but the result isn't too bad.

### CODE REVIEW

File: runner/run.go [hunk 1]
```diff
@@ -1,80 +1,206 @@
...
+	}
+
+	err = checkProviderResults(&run)
+	if err != nil {
+		return
+	}
+
+	// fetch secrets
+
+	if !runCached && (runnercfg.GetSecrets == nil || *runnercfg.GetSecrets) {
```
Comment (milescrabill) [inline runner/run.go]: Out of curiosity, why is this a `*bool`? Is it optional in the config?

File: runner/usage.go [hunk 2]
```diff
@@ -34,19 +19,14 @@ Configuration for this process is in the form of a YAML file with the following
...
-	if err != nil {
-		return nil, err
-	}
-	var runnercfg RunnerConfig
-	err = yaml.Unmarshal(data, &runnercfg)
-	if err != nil {
-		return nil, err
-	}
-	return &runnercfg, nil
+	getSecrets: if true (the default), then configuration is fetched from the secrets service
```
Comment (milescrabill) [inline runner/usage.go]: Can't we just supply JSON/YAML defaults via tags vs. making this a pointer, defaulting nil, and treating as true if nil?

File: credexp/credexp.go [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+
+	// the protocol (set in SetProtocol)
+	proto *protocol.Protocol
+
+	// a timer to handle sending a graceful-termination request before
+	// the credentials expire
+	credsExpireTimer *time.Timer
+}
+
+func New(run *run.Run) *CredExp {
```
[the remark this answers is not in the document]
Comment (djmitche) [reply credexp/credexp.go]: I did this to try to keep the same pattern we follow for other objects within the main `Run` function: set everything up, *then* start the worker, then set the protocol.  It avoids anything trying to use the protocol before the worker is ready.  In this case, I think we could create the CredExp after starting the worker, but I kind of like the regularity of calling `<thing>.SetProtocol(proto)` for all of the things.  What do you think?

[the remark this answers is not in the document]
Comment (milescrabill) [reply credexp/credexp.go]: I think that's fine, works for me. It just requires that you call `SetProtocol()` before using a `run.Run`.

File: runner/run.go [hunk 4]
```diff
@@ -1,80 +1,206 @@
...
-
-	// the protocol (set in SetProtocol)
-	proto *protocol.Protocol
-
-	// a timer to handle sending a graceful-termination request before
-	// the credentials expire
-	credsExpireTimer *time.Timer
-}
+// Check that the provider filled the run fields it was expected to.
+func checkProviderResults(run *run.Run) error {
```
[the remark this answers is not in the document]
Comment (djmitche) [reply runner/run.go]: I hadn't thought of that -- I was thinking of it as a utility function for the run..

File: runner/usage.go (hunk 2 above)
Comment (djmitche) [reply runner/usage.go]: golang says you shouldn't want to do that, so it's not supported :)

That said, I could apply the default in LoadRunnerConfig.

File: credexp/credexp.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (djmitche) [reply credexp/credexp.go]: I've been toying with the idea of having some kind of list of plugins that all match various interfaces, then something like

```go
for plugin := range plugins {
  sp, ok := plugin.(ProtoSetter)
  if ok {
    sp.SetProtocol(proto)
  }
}
```

and similarly for `ConfigureRun`, `WorkerStarted`, etc.

Maybe this is premature refactoring, I'm not sure.  If you'd like me to make this change I'm happy to, but my gut wants to leave it as-is.

[the remark this answers is not in the document]
Comment (djmitche) [reply credexp/credexp.go]: @milescrabill I'm waiting for your response on this..

[the remark this answers is not in the document]
Comment (djmitche) [reply credexp/credexp.go]: Oh, I somehow missed your reply, lol.. will land.

**Pull request merged**